crates/adapters/src/transport/pubsub/test.rs

```diff
@@ -0,0 +1,518 @@
+        config:
+            subscription: foo
+            emulator: "not_an_emulator:5678"
+    format:
+        name: csv
+    "#
+    );
+
+    match mock_input_pipeline::<TestStruct, TestStruct>(
+        serde_yaml::from_str(&config_str).unwrap(),
+        Relation::empty(),
+    ) {
+        Ok(_) => panic!("expected an error"),
+        Err(e) => println!("test_pubsub_input: Error (expected): {e}"),
+    };
+
+    info!("test_pubsub_input: Test: Specify invalid subscription name");
+
+    let config_str = format!(
+        r#"
+    stream: test_input
+    transport:
+        name: pub_sub_input
+        config:
+            project_id: {EMULATOR_PROJECT_ID}
+            emulator: "{emulator}"
+            subscription: this_subscription_does_not_exist
+    format:
+        name: csv
+    "#
+    );
+
+    match mock_input_pipeline::<TestStruct, TestStruct>(
+        serde_yaml::from_str(&config_str).unwrap(),
+        Relation::empty(),
+    ) {
+        Ok(_) => panic!("expected an error"),
+        Err(e) => println!("test_pubsub_input: Error (expected): {e}"),
+    };
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(2))]
+
+    // Running this test requires starting a Pub/Sub emulator:
+    // `gcloud beta emulators pubsub start --project=feldera-test --host-port=127.0.0.1:8685`.
+    // If you use a different port number, set the PUBSUB_EMULATOR_HOST env variable to
+    // point to that port.
+    #[cfg(feature = "pubsub-emulator-test")]
+    #[test]
+    fn proptest_pubsub_input_with_emulator(data in crate::test::generate_test_batches(0, 10, 100)) {
+        let emulator =
+            std::env::var("PUBSUB_EMULATOR_HOST").unwrap_or_else(|_| "127.0.0.1:8685".to_string());
+
+        let config  = [
+            ("project_id", EMULATOR_PROJECT_ID),
+            ("emulator", &emulator),
+        ];
+
+        test_pubsub_input(data, "test_pubsub_input", config.as_slice(), Some(&emulator));
+    }
+}
+
+proptest! {
+    // #![proptest_config(ProptestConfig::with_cases(2))]
+    #![proptest_config(ProptestConfig {
+        failure_persistence: None, // Disable persistence of failures
+        max_shrink_iters: 0,       // Disable shrinking
+        cases: 2,
+        .. ProptestConfig::default()
+    })]
+
+    // To run these tests, you need a Google Cloud account containing a project named 'feldera-test'.
+    // Authenticate to that account using 'gcloud' like this:
+    // `./bin/gcloud auth application-default login``
+    // Detailed instructions:
+    // https://cloud.google.com/pubsub/docs/authentication#client-libs
+
+    #[cfg(feature = "pubsub-gcp-test")]
+    #[test]
+    fn proptest_pubsub_input_with_gcp_auth(data in crate::test::generate_test_batches(0, 10, 100), topic_id in 0..1000000) {
+        crate::ensure_default_crypto_provider();
+
+        // Use randomized topic name; otherwise deleting and creating a topic is slow (~30s for me).
+        let topic = format!("test_pubsub_input{topic_id}");
```

Comment:
what is it with these cloud systems?

Reply:
it's geo-replicated and all, so takes time to get to consistent state. It's clearly not optimized for this scenarion